### ISSUE TITLE
Potential bug with 3d images/subresource layouts

### ISSUE THREAD

**Issue by RCalocaO** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2408]
I'm getting what I'm pretty sure are buggy validation errors when using a 3D texture aliased as 2d arrays. Attached is the API dump file, and in the companion log file, you can see the validation errors. This is with regards to 3d textures 0x961, 0x96c, 0x966, 0x96a, 0x963, etc at the bottom of the log file 
_(eg *** [ERROR:DS6] Obj 0x00000123314B0C00 Loc 2733 Cannot submit cmd buffer using image (0x966) [sub-resource: aspectMask 0x1 array layer 38, mip level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.._

Basically there's no way to change the layout on subslices of a 3d texture, and I had existing layout changes for the whole texture, so why is the validation complaining about those?

[vk_apidump2.zip](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/files/1721433/vk_apidump2.zip)

**Comment by RCalocaO** [comment]
[vk_apidump2.zip](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/files/1721432/vk_apidump2.zip)
Api dump file.

**Comment by daveh-lunarg** [comment]
Yeah, looks like another variant of issue #2210. Haven't yet been able to repro here, but the api dump should help with that.

@RCalocaO, what version of the SDK are you using?

**Comment by RCalocaO** [comment]
1.0.68.0

**Comment by daveh-lunarg** [comment]
@RCalocaO, any chance you can submit a small repro case?  Would be much faster than working through the api dump.

**Comment by RCalocaO** [comment]
I can maybe try to get a vktrace, would that be helpful?

**Comment by daveh-lunarg** [comment]
Sure, that'd help. Single frame to keep the size down.

If you've got an easy way to share it, just put the link here.  Otherwise I can set up a folder on share.lunarg.com to upload.

Thanks!

**Comment by RCalocaO** [comment]
https://www.dropbox.com/s/9sr45ptjfwueimt/vktrace_out.zip?dl=0

**Comment by RCalocaO** [comment]
I tried opening it with vktraceviewer, but it thought it was made with 1.0.65.1 for some reason (even though the engine loaded the C:\VulkanSDK\1.0.68.0\Source\lib\VkLayer_vktrace_layer.dll file).

**Comment by RCalocaO** [comment]
I have a packaged build I can send you, but it's 192MB.

**Comment by daveh-lunarg** [comment]
Awesome!

**Comment by davidlunarg** [comment]
@RCalocaO wrote:

> I tried opening it with vktraceviewer, but it thought it was made with 1.0.65.1 for some reason (even though the engine loaded the C:\VulkanSDK\1.0.68.0\Source\lib\VkLayer_vktrace_layer.dll file and I ran c:\VulkanSDK\1.0.68.0\bin\vktrace.exe)

I would like to investigate this. How did vktraceviewer identify that it thought the trace file was made with 1.0.65.1?

Also, you should be using VKLayer_vktrace_layer.dll in C:\VulkanSDK\1.0.68.0\Bin, not the one in Source\lib. See C:\VulkanSDK\1.0.68.0\Source\README.txt. But the dll in Source\lib should still work.

**Comment by RCalocaO** [comment]
@davidlunarg Not sure tbh why it would've pick up the wrong one. I'm always using the Source/lib versions as I want to see the source when validation fails (ie I have VK_LAYER_PATH=C:\VulkanSDK\1.0.68.0\Source\lib).

**Comment by daveh-lunarg** [comment]
Hi, @RCalocaO ,

I just pushed a branch (dave-gh2048) that eliminates the _"Cannot submit cmd buffer using image..."_ errors in your test case trace, by recording a layout transition for each depth slice in a 3D image whenever an image which uses the VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT_KHR flag is transitioned. 

If you're able to pull it and verify the fix it would be helpful. I'm sure it solves the particular instances in your trace file, but not 100% certain there aren't any remaining one-layout-per-image assumptions lurking in that code.
